### PR TITLE
Implements run and states for x86 MacOS

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**

### CODE REVIEW

**Comment by SuchAFuriousDeath** [comment]
Not sure what's up with the Windows build

**Comment by SuchAFuriousDeath** [comment]
It seems that we can't access the EFER register, but I assume the features we are interested in are already turned on?.

**Comment by SuchAFuriousDeath** [comment]
Not sure what to do about set_cs :(

**Comment by SuchAFuriousDeath** [comment]
I assume we're interested in ARM Linux and ARM Windows as well?

**Comment by ultimaweapon** [comment]
Seems like the Hypervisor Framework on macOS is the worst. A lot of restrictions.

> I assume we're interested in ARM Linux and ARM Windows as well?

If someone wiling to work on this.

**Comment by ultimaweapon** [comment]
BTW I'm surprised that WHP has a very good API.

**Comment by VocalFan** [comment]
> BTW I'm surprised that WHP has a very good API.

Microsoft takes pride in Hyper-V, so... Makes sense.

**Comment by SuchAFuriousDeath** [comment]
I think the segment registers can be set and read with this:
https://developer.apple.com/documentation/hypervisor/1441144-hv_vmx_vcpu_write_vmcs?language=objc

Here is a list of field ids. The segment registers do appear more than once. For example, for cs there's, VMCS_GUEST_CS, VMCS_HOST_CS, VMCS_GUEST_CS_LIMIT and VMCS_GUEST_CS_AR

https://developer.apple.com/documentation/hypervisor/1469436-virtual_machine_control_structur?language=objc

**Comment by ultimaweapon** [comment]
Should be `VMCS_GUEST_CS`. Not sure what the purpose of the other ones.

**Comment by SuchAFuriousDeath** [comment]
I guess the contents are the same as .Segment.Anonymous.Attributes on Windows? :eyes:

**Comment by ultimaweapon** [comment]
> I guess the contents are the same as .Segment.Anonymous.Attributes on Windows? 👀

This really implementation specific. Maybe finding some example on the internet should be easier if the docs does not help?

**Comment by ultimaweapon** [comment]
Just found a very good example: https://github.com/utshina/AHF-simple

**Comment by SuchAFuriousDeath** [comment]
A good example indeed. I managed to implemente exit_reason for x86 mac because of it

**Comment by SuchAFuriousDeath** [comment]
By the way the vm stuff seems to work correctly for now on my Linux machine.

**Comment by ultimaweapon** [comment]
No.

**Comment by SuchAFuriousDeath** [comment]
Oh, Where does it stop for you?

**Comment by ultimaweapon** [comment]
It is never stop because of `hlt`. In other words, it work as expected. Did you try on the latest `main` of from your branch?

**Comment by SuchAFuriousDeath** [comment]
Both, In main, I panic here 
https://github.com/obhq/obliteration/blob/573eca32403983bd2b4c6cafda9ed9630a3362b5/src/core/src/vmm/mod.rs#L615-L617
In my branch, i the same things happens in a different place, but it is equivalent.

**Comment by ultimaweapon** [comment]
The problem may be caused by a different Linux version or different CPU. What is your Linux version and CPU model? I was tested this with Arch on Ryzen 5600G.

**Comment by SuchAFuriousDeath** [comment]
EndeavourOS (Which is Arch-based), kernel version 6.9.7, cpu 11th Gen Intel(R) Core(TM) i5-11400H @ 2.70GHz

**Comment by ultimaweapon** [comment]
I suspect this #905 or https://github.com/firecracker-microvm/firecracker/issues/668 if your Linux running inside a VM.

**Comment by SuchAFuriousDeath** [comment]
I completely forgot that the issue applies to me :joy:  That might very well be it.

**Comment by SuchAFuriousDeath** [comment]
No nested VMs for me though.

**Comment by SuchAFuriousDeath** [comment]
> 

It doesn't seem to be it

**Comment by ultimaweapon** [comment]
Strange. At least on Windows with Ryzen it is working according to https://github.com/obhq/obliteration/pull/916#issue-2443300798

**Comment by ultimaweapon** [comment]
Can we ask another people who are using Linux with AMD to test this? If it is working I'm pretty sure it is because of #905.

**Comment by SuchAFuriousDeath** [comment]
What address should I set fr the TSS?

**Comment by ultimaweapon** [comment]
I don't know but any virtual address need to be mapped to physical address via page table otherwise the CPU will trigger page fault, which we currently does not setup a handler to handle it.

**Comment by SuchAFuriousDeath** [comment]
But the TSS that we have to set is a physical address, no?

**Comment by ultimaweapon** [comment]
I don't know. You should take a look on Intel manual. But AFAIK the only place that need to be a physical address is page tables and `cr3`.

**Comment by SuchAFuriousDeath** [comment]
image
It seems that I can get some information about the exit

**Comment by ultimaweapon** [comment]
BTW you should not make this PR contain too many things.

**Comment by SuchAFuriousDeath** [comment]
The exception is 0x80000021, the error code is 0.

**Comment by ultimaweapon** [comment]
We are not alone: https://forum.proxmox.com/threads/vm-shutdown-kvm-entry-failed-hardware-error-0x80000021.109410/

**Comment by ultimaweapon** [comment]
In our case I suspect TSS because `0x80000021` mean invalid guest state.

File: src/core/src/vmm/cpu.rs [hunk 1]
```diff
@@ -56,19 +56,23 @@ pub trait CpuStates {
...
 }
 
-/// Contains information when a VM exited because of I/O instructions.
-#[cfg(target_arch = "x86_64")]
-pub enum CpuIo<'a> {
-    Out(u16, &'a [u8]),
+#[derive(Debug)]
+pub enum ExitReason<'a> {
+    Hlt,
+    Other,
```
Comment (ultimaweapon) [inline src/core/src/vmm/cpu.rs:76]: We we need this?

File: src/core/src/vmm/mod.rs [hunk 2]
```diff
@@ -607,16 +607,14 @@ fn run_cpu(mut cpu: impl Cpu, args: &CpuArgs) {
...
+        match exit.reason() {
+            ExitReason::Hlt => {}
+            ExitReason::IoOut(0, data) => {
+                logs.extend_from_slice(data);
+                parse_logs(&args.logs, &mut logs);
             }
-        } else if !exit.is_hlt() {
-            todo!()
+            ExitReason::IoOut(_, _) => todo!(),
+            reason => todo!("unhandled VM exit reason: {reason:?}"),
```
Comment (ultimaweapon) [inline src/core/src/vmm/mod.rs:617]: This should not needed here. It should already handled by CPU implementation.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/core/src/vmm/cpu.rs (hunk 1 above)
Comment (SuchAFuriousDeath) [reply src/core/src/vmm/cpu.rs:76]: I'll revert it

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**